### ISSUE TITLE
[Bug]: Unable to use Deno.openKv()::enqueue() for production deploys

### ISSUE THREAD

**Issue by CatEars** [denoland/deploy_feedback#773]
### Problem description

I just deployed an application on the free tier that utilizes the `enqueue()` function of Deno KV. Similar to this code:

```typescript
const db = await Deno.openKv()
db.enqueue({ some: "data" })
```

This works completely fine in my local dev environment, but when I push to Deno Deploy (and deploy to production!) I get the following message in the logs:

![image](https://github.com/user-attachments/assets/7ff265c1-112c-4496-b2bb-3e3d97ee68d1)

### Steps to reproduce

1. Create application utilizing `enqueue()` and `listenQueue()` of `Deno.openKv()`
2. Deploy to production on Deno Deploy free tier using `deployctl deploy`
3. Promote deployment to production deploy
4. Trigger `enqueue()` by interacting with the app

### Expected behavior

Messages that are enqueued for the production app should not be dropped

### Environment

- Deno Deploy (gcp-europe-west3)
- deployctl 1.13.0

**Comment by magurotuna** [comment]
Hi @CatEars, I would like to clarify a few points for investigation:

- Can you share your Deno Deploy project name where you encounter this issue?
- How did you promote the deployment to production? Via dashboard, or deployctl command?

**Comment by CatEars** [comment]
Hi @magurotuna. The project name is "yeet". I believe I have promoted to production via the website / project homepage.

**Comment by magurotuna** [comment]
@CatEars Thanks for the info. We will investigate it.

**Comment by magurotuna** [comment]
I've looked into it and discovered that queue handlers are not properly set when deployments are promoted to production. The hidden assumption is that in order for a preview deployment to get promoted to prod, it must have been a prod deployment before - currently there is no mechanism to enforce this.

So the intended way to _effectively_ promote a preview deployment (that has never been a prod deployment) to prod is just _redeploy_ it with `deployctl deploy --prod` instead of promoting it to prod via Web UI.

**Comment by CatEars** [comment]
Thanks for a thorough explanation. I will try to get past the "hidden" assumption with the info you provided. Hopefully in the evening today ;)

**Comment by pazaan** [comment]
Running `deployctl deploy --prod` solved the issue for me (but `deployctl redeploy --prod` did not).

**Comment by magurotuna** [comment]
Great, glad it's solved :)

**Comment by ChristianSiegert** [comment]
> Great, glad it's solved :)

It’s not fully solved. If promoting a preview deployment to production deployment does not work correctly, promoting should not be possible. In this case promotion will lead to data loss – enqueued messages are being silently dropped (except for a log entry that only warns after the fact).

**Comment by magurotuna** [comment]
Yes, that's right. We are now working on adding a validation logic to prevent preview deployments that have _never_ been production deployments from being promoted to prod.

**Comment by CatEars** [comment]
I have now run `deployctl deploy --prod` and can confirm it worked. Thanks for the help @magurotuna 🫸🫷
